EXTENSIONS.md

```diff
@@ -0,0 +1,236 @@
+
+## Function decorators and annotations
+
+The `@logged` decorator from earlier would Just Work(TM) on a function, with function decorators!
+
+```js
+@logged
+@{x: "y"}
+function f() { }
+
+f();                        // prints logging information
+f[Symbol.annotations][0].x  // "y"
+```
+
+Function declarations with decorators or annotations are not hoisted. This is because it would be unintuitive to reorder the evaluation of the decorator or annotation expressions.
+
+Instead, functions with decorators or annotations are defined only when their declaration is reached. If they are used before they are defined, a ReferenceError is thrown, like classes. This ReferenceError condition is sometimes referred to as a "temporal dead zone" (TDZ). The TDZ risks unfortunate situations when refactoring, but at least those situations lead to easy-to-debug errors rather than the wrong function being run.
+
+Function decorator details:
+- First parameter: the function being decorated (or, whatever the next inner decorator returned)
+- Second parameter: a context object which just has `{ kind: 'function' }`
+- Return value: a new function, or undefined to keep the same function
+
+The inner binding of a function expression inside itself is in TDZ until all the function decorators run.
+
+## Parameter decorators and annotations
+
+A parameter decorator wraps the value of a function/method argument. It returns a function which does the wrapping.
+
+```js
+function dec(_, context) {
+  assert(context.kind === "parameter");
+  return arg => arg - 1;
+}
+
+function f(@dec @{x: "y"} arg) { return arg * 2 ; }
+
+f(5)  // 8
+f[Symbol.annotations].parameters[0].x  // "y"
+```
+
+Functions with parameters that are decorated or annotated are treated similarly to decorated/annotated functions: they are not hoisted, and are in TDZ until their definition is executed.
+
+Parameter decorator details:
+- First parameter: undefined
+- Second parameter: a context object which just has `{ kind: 'parameter' }`
+- Return value: a function which takes a parameter value and returns a new parameter value. The function is called with the `this` value that the surrounding function is called with.
+
+This example can be desugared as
+
+```js
+let decInstance = dec(undefined, {kind: "parameter"});
+
+function f(arg_) {
+  const arg = decInstance(arg_);
+  return arg * 2 ;
+}
+
+f[Symbol.annotations] = {}
+f[Symbol.annotations].parameters = []
+f[Symbol.annotations].parameters[0] = {x: "y"};
+```
+
+## `let` decorators
+
+Variables declared with `let` can be decorated, converting them into special getter/setter pairs that are invoked when the variable is read or written.
+
+```js
+let @deprecated x = 1;
+++x;  // Shows deprecation warnings for read and write
+```
+
+`let` decorators might be useful for systems using reactivity based on local variables, e.g. "hooks" systems.
```

Comment:
```js
function FutureReactComponent() {
    let @state count = 1
    const @callback onClick = () => count++
    const @ref anchor = null
    ; @effect(count) {
        const callback = () => console.log(count)
        document.addEventListener('click', callback)
        ; () => document.removeEventListener('click', callback)
    };
    return <a ref={anchor} onClick={onClick}>{count}</a>
}
```

Reply:
As you can see in the `@effect(...dependencies) { block }`, the last expression is returned to the decorator based on my understanding of the block decorator. Is that somehow like `do expression`?

Reply:
I don't see how it could work to list the dependencies as decorator parameters, as the parameters are just evaluated as JS expressions.

Reply:
It's pattern of React today. `useEffect(fn, [count])`

Reply:
Oh, yes, this does seem like a good idea for an application in React! In other reactivity models which are not based on running the whole thing each frame, another phrasing would likely be needed, which is why I got confused.

Reply:
I wanted to mention: I was picturing that decorated blocks wouldn't suddenly get to return a value. Maybe we actually want a decorated do expression or arrow function for this case... Also, I'm not sure if the ref usage would work, since you need to pass the ref itself into the prop, not the result of getting the ref.

Reply:
Would also like to point out it'd give Svelte an opportunity to avoid some of its hackiness as well.

Here's one component [ripped from its examples](https://svelte.dev/examples#hacker-news) as an example:

```svelte
<script>
	import Summary from "./Summary.svelte";

	const PAGE_SIZE = 20;

	export let page;

	let items;
	let offset;

	$: fetch(`https://node-hnapi.herokuapp.com/news?page=${page}`)
		.then(r => r.json())
		.then(data => {
			items = data;
			offset = PAGE_SIZE * (page - 1);
			window.scrollTo(0, 0);
		});
</script>

<style>
	a {
		padding: 2em;
		display: block;
	}

	.loading {
		opacity: 0;
		animation: 0.4s 0.8s forwards fade-in;
	}

	@keyframes fade-in {
		from { opacity: 0; }
		to { opacity: 1; }
	}
</style>

{#if items}
	{#each items as item, i}
		<Summary {item} {i} {offset}/>
	{/each}

	<a href="#/top/{page + 1}">page {page + 1}</a>
{:else}
	<p class="loading">loading...</p>
{/if}
```

If they switched to some of the proposed extensions, it might look closer to this:

```svelte
<script>
	import { prop, react } from "svelte";
	import Summary from "./Summary.svelte";

	const PAGE_SIZE = 20;

	let @prop page;

	let items;
	let offset;

	@react {
		fetch(`https://node-hnapi.herokuapp.com/news?page=${page}`)
			.then(r => r.json())
			.then(data => {
				items = data;
				offset = PAGE_SIZE * (page - 1);
				window.scrollTo(0, 0);
			});
	}
</script>

<style>
	a {
		padding: 2em;
		display: block;
	}

	.loading {
		opacity: 0;
		animation: 0.4s 0.8s forwards fade-in;
	}

	@keyframes fade-in {
		from { opacity: 0; }
		to { opacity: 1; }
	}
</style>

{#if items}
	{#each items as item, i}
		<Summary {item} {i} {offset}/>
	{/each}

	<a href="#/top/{page + 1}">page {page + 1}</a>
{:else}
	<p class="loading">loading...</p>
{/if}
```